src/main/kotlin/org/rust/ide/ssr/RsMatchingVisitor.kt

```diff
@@ -0,0 +1,245 @@
+        matchingVisitor.result = matchingVisitor.matchSequentially(o.tupleFieldDeclList, fields.tupleFieldDeclList)
+    }
+
+    override fun visitTupleFieldDecl(o: RsTupleFieldDecl) {
+        val field = getElement<RsTupleFieldDecl>() ?: return
+        matchingVisitor.result =
+            matchOuterAttrList(o, field) &&
+                match(o.vis, field.vis) &&
+                match(o.typeReference, field.typeReference)
+    }
+
+    override fun visitTypeReference(o: RsTypeReference) {
+        val typeReference = getElement<RsTypeReference>() ?: return
+        // TODO: implement individual type references
+        matchingVisitor.result = matchTextOrVariable(o, typeReference)
+    }
+
+    override fun visitRefLikeType(o: RsRefLikeType) {
+        val refType = getElement<RsRefLikeType>() ?: return
+        matchingVisitor.result = match(o.typeReference, refType.typeReference) &&
+            match(o.lifetime, refType.lifetime) &&
+            matchLeaf(o.mut, refType.mut) &&
+            matchLeaf(o.const, refType.const) &&
+            matchLeaf(o.mul, refType.mul) &&
+            matchLeaf(o.and, refType.and)
+    }
+
+    override fun visitLifetime(o: RsLifetime) {
+        val lifetime = getElement<RsLifetime>() ?: return
+        matchingVisitor.result = matchTextOrVariable(o, lifetime)
+    }
+
+    override fun visitVis(o: RsVis) {
+        val vis = getElement<RsVis>() ?: return
+        matchingVisitor.result = matchTextOrVariable(o, vis)
+    }
+
+    override fun visitInnerAttr(o: RsInnerAttr) {
+        val attr = getElement<RsInnerAttr>() ?: return
+        matchingVisitor.result = match(o.metaItem, attr.metaItem)
+    }
+
+    override fun visitOuterAttr(o: RsOuterAttr) {
+        val attr = getElement<RsOuterAttr>() ?: return
+        matchingVisitor.result = match(o.metaItem, attr.metaItem)
+    }
+
+    override fun visitMetaItem(o: RsMetaItem) {
+        val metaItem = getElement<RsMetaItem>() ?: return
+        matchingVisitor.result = match(o.compactTT, metaItem.compactTT) &&
+            match(o.litExpr, metaItem.litExpr) &&
+            match(o.metaItemArgs, metaItem.metaItemArgs) &&
+            match(o.path, metaItem.path)
+    }
+
+    override fun visitMetaItemArgs(o: RsMetaItemArgs) {
+        val metaItemArgs = getElement<RsMetaItemArgs>() ?: return
+        matchingVisitor.result = matchingVisitor.matchInAnyOrder(o.metaItemList, metaItemArgs.metaItemList)
+    }
+
+    override fun visitPathType(o: RsPathType) {
+        val path = getElement<RsPathType>() ?: return
+        matchingVisitor.result = match(o.path, path.path)
+    }
+
+    override fun visitPath(o: RsPath) {
+        val path = getElement<RsPath>() ?: return
+
+        matchingVisitor.result = matchIdentifier(o.identifier, path.identifier) &&
+            match(o.typeArgumentList, path.typeArgumentList) &&
+            match(o.valueParameterList, path.valueParameterList) &&
+            match(o.path, path.path)
```

Comment:
Let's also match `o.coloncolon` to distinguish between `foo::bar` and `::foo::bar`

Reply:
This causes problems (`Foo` doesn't match `x::Foo` anymore), let's postpone and include this change in the revamp of path matching.